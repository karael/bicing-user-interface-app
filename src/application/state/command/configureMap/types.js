import {
  FAILURE, START, SUCCESS,
} from 'application/state/stateConstants';
import { defineAction } from 'redux-define';

export const CONFIGURE = defineAction(
  'CONFIGURE',
  [START, SUCCESS, FAILURE],
  defineAction('command/configureMap'),
);

export default CONFIGURE;
