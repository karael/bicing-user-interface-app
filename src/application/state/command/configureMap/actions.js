import { CONFIGURE } from 'application/state/command/configureMap/types';

export const configureStart = (latitude, longitude, limit) => ({
  error: false,
  payload: { latitude, longitude, limit },
  type: CONFIGURE.START,
});

export const configureSuccess = data => ({
  error: false,
  payload: data,
  type: CONFIGURE.SUCCESS,
});

export const configureFailure = error => ({
  error: true,
  payload: error,
  type: CONFIGURE.FAILURE,
});
