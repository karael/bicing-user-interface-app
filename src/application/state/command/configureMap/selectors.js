const selector = state => state.command.configureMap;

export const latitude = state => selector(state).latitude;
export const longitude = state => selector(state).longitude;
export const limit = state => selector(state).limit;
export const data = state => selector(state).data;
export const zoom = state => data(state);
export const error = state => selector(state).error;
