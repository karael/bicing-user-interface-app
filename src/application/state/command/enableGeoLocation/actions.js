import { ENABLE } from 'application/state/command/enableGeoLocation/types';

export const enable = itineraryStep => ({
  payload: { itineraryStep },
  type: ENABLE.START,
});

export default enable();
