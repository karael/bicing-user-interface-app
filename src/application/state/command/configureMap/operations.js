import { configureSuccess, configureFailure } from 'application/state/command/configureMap/actions';
import { CONFIGURE } from 'application/state/command/configureMap/types';
import { put, takeLatest } from 'redux-saga/effects';
import ByGeoLocationFilter from 'application/state/filter/ByGeoLocationFilter';

export function* configure(action) {
  try {
    const { latitude, longitude, limit } = action.payload;
    const filter = ByGeoLocationFilter.fromRawValues(latitude, longitude, limit);

    yield put(configureSuccess(filter.zoom()));
  } catch (exception) {
    yield put(configureFailure(exception));
  }
}

export default function* operation() {
  yield takeLatest(CONFIGURE.START, configure);
}
