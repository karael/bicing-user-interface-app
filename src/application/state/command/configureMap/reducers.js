import { CONFIGURE } from 'application/state/command/configureMap/types';
import produce from 'immer';
import { createReducer } from 'reduxsauce';

export const INITIAL_STATE = {
  error: false,
  data: undefined,
  latitude: undefined,
  longitude: undefined,
  limit: undefined,
};

export const configureStart = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.latitude = action.payload.latitude;
  draft.longitude = action.payload.longitude;
  draft.limit = action.payload.limit;
});

export const configureSuccess = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = action.payload;
});

export const configureFailure = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.error = true;
  draft.data = action.payload;
});

export const HANDLERS = {
  [CONFIGURE.START]: configureStart,
  [CONFIGURE.SUCCESS]: configureSuccess,
  [CONFIGURE.FAILURE]: configureFailure,
};

export default createReducer(INITIAL_STATE, HANDLERS);
