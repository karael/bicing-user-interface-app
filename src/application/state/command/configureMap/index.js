import * as actions from 'application/state/command/configureMap/actions';
import operation from 'application/state/command/configureMap/operations';
import reducer from 'application/state/command/configureMap/reducers';
import * as types from 'application/state/command/configureMap/types';
import * as selectors from 'application/state/command/configureMap/selectors';

export {
  actions,
  operation,
  selectors,
  types,
};

export default reducer;
