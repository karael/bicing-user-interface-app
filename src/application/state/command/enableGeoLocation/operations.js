import { ENABLE } from 'application/state/command/enableGeoLocation/types';
import { takeLatest } from 'redux-saga/effects';

export function* enable() {
  yield null;
}

export default function* operation() {
  yield takeLatest(ENABLE.START, enable);
}
