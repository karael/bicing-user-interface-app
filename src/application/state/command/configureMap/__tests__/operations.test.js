import * as actions from 'application/state/command/configureMap/actions';
import operation, { configure } from 'application/state/command/configureMap/operations';
import { CONFIGURE } from 'application/state/command/configureMap/types';
import { expectSaga, testSaga } from 'redux-saga-test-plan';
import { fetch } from 'application/state/query/stations/operations';

describe('application/state/command/configureMap/operations', () => {
  test('should wait for a configure start event to configure map with operation()', () => {
    testSaga(operation)
      .next()
      .takeLatestEffect(CONFIGURE.START, configure);
  });

  test('should notify success configuration with configure() generator', () => {
    const action = {
      error: false,
      payload: { latitude: 41.356, longitude: 2.1234, limit: 200 },
      type: CONFIGURE.START,
    };
    const expectedConfiguration = 12;

    expectSaga(fetch, action)
      .put(actions.configureSuccess(expectedConfiguration))
      .run();
  });


  test('should handle error when byGeoLocationFilter creation is not validated', () => {
    const action = {
      error: false,
      payload: { latitude: 'bad_value', longitude: 'bad_value', limit: 'bad_value' },
      type: CONFIGURE.START,
    };

    return expectSaga(fetch, action)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
