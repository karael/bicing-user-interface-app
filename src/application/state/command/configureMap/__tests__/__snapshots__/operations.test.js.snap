// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/command/configureMap/operations should handle error when byGeoLocationFilter creation is not validated 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/stations/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": Object {
              "latitude": "bad_value",
              "limit": "bad_value",
              "longitude": "bad_value",
            },
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": "latitude",
                  "label": "latitude",
                  "value": "bad_value",
                },
                "message": "\\"latitude\\" must be a number",
                "path": Array [
                  "latitude",
                ],
                "type": "number.base",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/stations/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;
